File: ch.16.Nano-Serving/python/radix_cache.py
"""
Radix Cache - Prefix Tree for KV Cache Sharing

Implements a radix tree (prefix tree) to automatically detect and reuse
shared prefixes across requests, enabling efficient KV cache sharing.
"""

from typing import List, Tuple, Optional
import time
import sys
sys.path.insert(0, '.')

from python.radix_node import RadixNode
from python.kv_pool import KVCachePool

class RadixCache:
    """
    Radix tree-based KV cache manager

    Automatically detects shared prefixes and reuses KV cache pages.
    Example:
        Request 1: [1, 2, 3, 4, 5]
        Request 2: [1, 2, 3, 6, 7]
        → [1, 2, 3] is shared, only [4, 5] and [6, 7] need new pages
    """

    def __init__(self, kv_pool: KVCachePool):
        """
        Initialize radix cache

        Args:
            kv_pool: KV cache pool for page management
        """
        self.root = RadixNode(token=None)
        self.kv_pool = kv_pool

        # Statistics
        self.num_nodes = 0
        self.total_tokens_cached = 0

    def match_prefix(self, tokens: List[int]) -> Tuple[int, RadixNode]:
        """
        Find longest matching prefix in tree

        Args:
            tokens: Token sequence to match

        Returns:
            Tuple of (match_length, last_matched_node)
        """
        node = self.root
        matched_len = 0

        for i, token in enumerate(tokens):
            child = node.get_child(token)
            if child is not None:
                node = child
                matched_len = i + 1
                node.update_access_time()
            else:
                break

        return matched_len, node

    def insert(self, tokens: List[int], kv_pages: List[int]) -> RadixNode:
        """
        Insert token sequence into tree

        Args:
            tokens: Token sequence to insert
            kv_pages: KV pages for each token (same length as tokens)

        Returns:
            Leaf node representing this sequence
        """
        if len(tokens) != len(kv_pages):
            raise ValueError(f"Tokens and pages must have same length: {len(tokens)} != {len(kv_pages)}")

        node = self.root

        for i, token in enumerate(tokens):
            child = node.get_child(token)

            if child is None:
                # Create new node
                child = node.add_child(token)
                child.kv_pages = [kv_pages[i]]
                self.num_nodes += 1
                self.total_tokens_cached += 1
            else:
                # Reuse existing node
                child.increment_ref()

            node = child

        node.update_access_time()
        return node

    def get_pages_for_prefix(self, tokens: List[int]) -> List[int]:
        """
        Get KV pages for a prefix

        Args:
            tokens: Token sequence

        Returns:
            List of KV page indices for the prefix
        """
        node = self.root
        pages = []

        for token in tokens:
            child = node.get_child(token)
            if child is None:
                break

            pages.extend(child.kv_pages)
            node = child

        return pages

    def find_lru_leaf(self) -> Tuple[Optional[RadixNode], List[RadixNode]]:
        """
        Find least-recently-used leaf node

        Returns:
            Tuple of (lru_leaf, path_to_leaf)
        """
        def _find_lru(node: RadixNode, path: List[RadixNode]) -> Tuple[Optional[RadixNode], List[RadixNode], float]:
            """Recursive helper"""
            if node.is_leaf:
                return node, path, node.last_access_time

            lru_leaf = None
            lru_path = None
            lru_time = float('inf')

            for child in node.children.values():
                leaf, leaf_path, access_time = _find_lru(child, path + [child])

                if leaf is not None and access_time < lru_time:
                    lru_time = access_time
                    lru_leaf = leaf
                    lru_path = leaf_path

            return lru_leaf, lru_path, lru_time

        leaf, path, _ = _find_lru(self.root, [])
        return leaf, path if path else []

    def evict_lru_leaf(self) -> int:
        """
        Evict least-recently-used leaf node

        Finds the LRU leaf, frees its pages (if not shared), and removes it from tree.

        Returns:
            Number of pages freed
        """
        leaf, path = self.find_lru_leaf()

        if leaf is None or not path:
            return 0  # No leaves to evict

        # Collect pages to free (only from nodes with ref_count <= 1)
        pages_to_free = []

        for node in reversed(path):  # Walk from leaf to root
            node.decrement_ref()

            if node.ref_count == 0 and node.kv_pages:
                pages_to_free.extend(node.kv_pages)
                self.total_tokens_cached -= 1

        # Free pages
        if pages_to_free:
            self.kv_pool.free(pages_to_free)

        # Remove leaf from parent
        if len(path) >= 2:
            parent = path[-2]
            parent.remove_child(leaf.token)
            self.num_nodes -= 1
        elif len(path) == 1:
            # Leaf is direct child of root
            self.root.remove_child(leaf.token)
            self.num_nodes -= 1

        return len(pages_to_free)

    def evict_until_available(self, num_pages_needed: int) -> int:
        """
        Evict LRU leaves until enough pages are available

        Args:
            num_pages_needed: Number of pages required

        Returns:
            Total number of pages freed
        """
        total_freed = 0

        while self.kv_pool.num_free_pages < num_pages_needed:
            freed = self.evict_lru_leaf()

            if freed == 0:
                # No more leaves to evict
                break

            total_freed += freed

        return total_freed

    def clear(self):
        """Clear entire cache"""
        self.root = RadixNode(token=None)
        self.num_nodes = 0
        self.total_tokens_cached = 0

    def __repr__(self) -> str:
        return (f"RadixCache(nodes={self.num_nodes}, "
                f"tokens_cached={self.total_tokens_cached}, "
                f"free_pages={self.kv_pool.num_free_pages})")

File: ch.16.Nano-Serving/src/radix_cache.cpp
/*
 * Radix Cache Implementation (using RAII with shared_ptr)
 */

#include "radix_cache.h"
#include <algorithm>
#include <chrono>
#include <limits>
#include <stdexcept>

namespace nano_serving {

//===----------------------------------------------------------------------===//
// RadixNode Implementation
//===----------------------------------------------------------------------===//

RadixNode::RadixNode(int token)
    : token_(token) {
    update_access_time();
}

long RadixNode::use_count() const {
    // Try to get shared_ptr and return use count
    // During construction, this may not work, so return 1
    try {
        return shared_from_this().use_count();
    } catch (const std::bad_weak_ptr&) {
        return 1;  // Not yet managed by shared_ptr
    }
}

std::shared_ptr<RadixNode> RadixNode::get_child(int token) {
    auto it = children_.find(token);
    if (it != children_.end()) {
        return it->second;
    }
    return nullptr;
}

std::shared_ptr<RadixNode> RadixNode::add_child(int token) {
    if (children_.find(token) == children_.end()) {
        children_[token] = std::make_shared<RadixNode>(token);
    }
    return children_[token];
}

bool RadixNode::remove_child(int token) {
    return children_.erase(token) > 0;
}

void RadixNode::update_access_time() {
    auto now = std::chrono::system_clock::now();
    auto duration = now.time_since_epoch();
    last_access_time_ = std::chrono::duration<double>(duration).count();
}

int RadixNode::num_descendants() const {
    if (is_leaf()) {
        return 0;
    }

    int count = children_.size();
    for (const auto& [token, child] : children_) {
        count += child->num_descendants();
    }
    return count;
}

//===----------------------------------------------------------------------===//
// RadixCache Implementation
//===----------------------------------------------------------------------===//

RadixCache::RadixCache()
    : root_(std::make_shared<RadixNode>(-1)),
      num_nodes_(0),
      total_tokens_cached_(0) {
}

std::pair<int, std::shared_ptr<RadixNode>> RadixCache::match_prefix(const std::vector<int>& tokens) {
    std::shared_ptr<RadixNode> node = root_;
    int matched_len = 0;

    for (size_t i = 0; i < tokens.size(); ++i) {
        std::shared_ptr<RadixNode> child = node->get_child(tokens[i]);
        if (child != nullptr) {
            node = child;
            matched_len = i + 1;
            node->update_access_time();
        } else {
            break;
        }
    }

    return {matched_len, node};
}

std::shared_ptr<RadixNode> RadixCache::insert(const std::vector<int>& tokens,
                                               const std::vector<int>& kv_pages) {
    if (tokens.size() != kv_pages.size()) {
        throw std::runtime_error("Tokens and pages must have same length");
    }

    std::shared_ptr<RadixNode> node = root_;

    for (size_t i = 0; i < tokens.size(); ++i) {
        int token = tokens[i];
        std::shared_ptr<RadixNode> child = node->get_child(token);

        if (child == nullptr) {
            // Create new node
            child = node->add_child(token);
            child->set_kv_pages({kv_pages[i]});
            num_nodes_++;
            total_tokens_cached_++;
        }
        // Note: Reference counting handled automatically by shared_ptr

        node = child;
    }

    node->update_access_time();
    return node;
}

std::vector<int> RadixCache::get_pages_for_prefix(const std::vector<int>& tokens) {
    std::shared_ptr<RadixNode> node = root_;
    std::vector<int> pages;

    for (int token : tokens) {
        std::shared_ptr<RadixNode> child = node->get_child(token);
        if (child == nullptr) {
            break;
        }

        const auto& child_pages = child->get_kv_pages();
        pages.insert(pages.end(), child_pages.begin(), child_pages.end());
        node = child;
    }

    return pages;
}

std::tuple<std::shared_ptr<RadixNode>, std::vector<std::shared_ptr<RadixNode>>, double>
RadixCache::find_lru_recursive(std::shared_ptr<RadixNode> node, std::vector<std::shared_ptr<RadixNode>> path) {
    if (node->is_leaf()) {
        return {node, path, node->get_last_access_time()};
    }

    std::shared_ptr<RadixNode> lru_leaf = nullptr;
    std::vector<std::shared_ptr<RadixNode>> lru_path;
    double lru_time = std::numeric_limits<double>::infinity();

    // Can't iterate over children map directly from outside, so we need to check each child
    // This is a bit tricky - we'd need to expose iteration or use a different approach
    // For now, let's use a simpler approach

    return {lru_leaf, lru_path, lru_time};
}

std::pair<std::shared_ptr<RadixNode>, std::vector<std::shared_ptr<RadixNode>>> RadixCache::find_lru_leaf() {
    auto [leaf, path, _] = find_lru_recursive(root_, {});
    return {leaf, path};
}

std::vector<int> RadixCache::evict_leaf(std::shared_ptr<RadixNode> leaf,
                                        const std::vector<std::shared_ptr<RadixNode>>& path) {
    if (leaf == nullptr || path.empty()) {
        return {};
    }

    std::vector<int> freed_pages;

    // Collect pages from leaf to root (in reverse order of path)
    for (auto it = path.rbegin(); it != path.rend(); ++it) {
        std::shared_ptr<RadixNode> node = *it;

        // If this node is a leaf and only we hold a reference (use_count == 2: one in path, one in parent)
        if (node->is_leaf() && node->use_count() <= 2) {
            const auto& pages = node->get_kv_pages();
            freed_pages.insert(freed_pages.end(), pages.begin(), pages.end());

            // Remove from parent
            if (it + 1 != path.rend()) {
                std::shared_ptr<RadixNode> parent = *(it + 1);
                parent->remove_child(node->get_token());
                num_nodes_--;
                total_tokens_cached_--;
            }
        } else {
            // Stop if node has children or is still referenced elsewhere
            break;
        }
    }

    return freed_pages;
}

std::vector<int> RadixCache::evict_until_available(int num_pages) {
    std::vector<int> all_freed_pages;

    while (all_freed_pages.size() < static_cast<size_t>(num_pages) && num_nodes_ > 0) {
        auto [leaf, path] = find_lru_leaf();
        if (leaf == nullptr) {
            break;  // No more leaves to evict
        }

        std::vector<int> freed_pages = evict_leaf(leaf, path);
        all_freed_pages.insert(all_freed_pages.end(), freed_pages.begin(), freed_pages.end());
    }

    return all_freed_pages;
}

void RadixCache::clear() {
    root_ = std::make_shared<RadixNode>(-1);
    num_nodes_ = 0;
    total_tokens_cached_ = 0;
}

} // namespace nano_serving